src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs

```diff
@@ -442,19 +441,125 @@ private BoundExpression VisitObjectInitializerMember(BoundObjectInitializerMembe
+                                        BoundPassByCopy { Expression: BoundLocal { LocalSymbol: SynthesizedLocal } }]
+                        });
+                    }
+                    else
+                    {
+                        transformedImplicitIndexer = TransformImplicitIndexerAccess(implicitIndexer,
+                            isRegularAssignmentOrRegularCompoundAssignment: true, inInitializer: true, result, temps, isDynamicAssignment: false);
+
+                        Debug.Assert(transformedImplicitIndexer is BoundArrayAccess { Indices: [BoundLocal { LocalSymbol: SynthesizedLocal }] }
+                            or BoundIndexerAccess { Arguments: [BoundLiteral or BoundLocal { LocalSymbol: SynthesizedLocal }] }
+                            or BoundLocal { LocalSymbol: SynthesizedLocal });
+                    }
+
+                    if (transformedImplicitIndexer is BoundIndexerAccess indexerAccess)
+                    {
+                        rewrittenAccess = VisitIndexerAccess(indexerAccess, isLeftOfAssignment: !isRhsNestedInitializer || indexerAccess.Indexer.RefKind != RefKind.None);
+                    }
+                    else
+                    {
+                        rewrittenAccess = transformedImplicitIndexer;
+                    }
+
+                    if (!isRhsNestedInitializer)
+                    {
+                        var rewrittenRight = VisitExpression(right);
+                        result.Add(MakeStaticAssignmentOperator(assignment.Syntax, rewrittenAccess, rewrittenRight, isRef: false, assignment.Type, used: false));
+                        return;
+                    }
+
+                    break;
+
                 default:
-                    throw ExceptionUtilities.UnexpectedValue((rewrittenLeft ?? assignment.Left).Kind);
+                    throw ExceptionUtilities.UnexpectedValue(left.Kind);
+            }
+
+            AddObjectOrCollectionInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenAccess, right);
+            return;
+
+            static bool onlyContainsEmptyLeafNestedInitializers(BoundExpression expression)
+            {
+                return expression is BoundObjectInitializerExpression initializer
+                    && initializer.Initializers.All(e => e is BoundAssignmentOperator assignment && onlyContainsEmptyLeafNestedInitializers(assignment.Right));
             }
 
-            AddObjectOrCollectionInitializers(ref dynamicSiteInitializers, ref temps, result, rewrittenAccess, assignment.Right);
+            void addIndexes(ArrayBuilder<BoundExpression> result, BoundAssignmentOperator assignment)
+            {
+                // If we have an element access of the form `[arguments] = { ... }`, we'll evaluate `arguments` only
+                var lhs = assignment.Left;
+                if (lhs is BoundObjectInitializerMember initializerMember)
```

Comment:
>if (lhs is BoundObjectInitializerMember initializerMember) [](http://example.com/codeflow?start=16&length=58)

I think this ```if/else``` sequence should explicitly handle everything that can come through this helper and have an explicit ```throw ExceptionUtilities.Unreachable()``` for everything else.